### PR TITLE
XMLRPC API ContentManagement createAppStreamFilters

### PR DESCRIPTION
**Pull request by wombelix**
## What does this PR change?

Adds API Endpoint to create AppStreamFilters based on Template similar as the CLM Web UI functionality.
Re-use Classes and methods also used for the Web UI Implementation: https://github.com/uyuni-project/uyuni/pull/4147

## GUI diff

No difference.

Before:

After:

- [x] **DONE**

## Documentation
- No documentation needed: API Documentation generated by javadoc

- [x] **DONE**

## Test coverage
- No tests: already covered in `java/code/src/com/redhat/rhn/manager/contentmgmt/test/ContentManagerTest.java`

- [x] **DONE**

## Links

none

- [x] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests" 
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

**Comment by wombelix** [comment]
So far so good, looking for someone who want's to Review the PR :)
I didn't implemented own tests because I'm re-using existing classes/methods that seem to already covered.

File: java/code/src/com/redhat/rhn/frontend/xmlrpc/contentmgmt/ContentManagementHandler.java [hunk 1]
```diff
@@ -645,6 +651,39 @@ public ContentFilter createFilter(User loggedInUser, String name, String rule, S
...
 
+    /**
+     * Create new {@link ContentFilter}s for all AppStream modules with default streams
+     *
+     * @param loggedInUser the logged in user
+     * @param prefix the filter name prefix
+     * @param channelLabel label of the modular channel
+     * @param projectLabel label of the Content Lifecycle Project
+     * @throws EntityExistsFaultException when Filter already exist
+     * @return List of created and successfully attached Filter
```
Comment (mcalmer) [inline java/code/src/com/redhat/rhn/frontend/xmlrpc/contentmgmt/ContentManagementHandler.java:662]: The API has additional doc strings to autogenerate the developers documentation.

**Review by mcalmer — CHANGES_REQUESTED** [review]

File: java/code/src/com/redhat/rhn/frontend/xmlrpc/contentmgmt/ContentManagementHandler.java (hunk 1 above)
Comment (cbbayburt) [reply java/code/src/com/redhat/rhn/frontend/xmlrpc/contentmgmt/ContentManagementHandler.java:662]: @wombelix we'll add a wiki guide for writing documentation for API methods in Javadoc soon. But until then, please use the existing examples to add a brief documentation to your method.

Basically, you'll need to fill `@xmlrpc.doc` for a description, `@xmlrpc.param` for each parameter with their names and types, and `@xmlrpc.returntype` for the return value.

File: java/code/src/com/redhat/rhn/frontend/xmlrpc/contentmgmt/ContentManagementHandler.java [hunk 2]
```diff
@@ -645,6 +651,39 @@ public ContentFilter createFilter(User loggedInUser, String name, String rule, S
...
+     */
+    public List<ContentFilter> createAppStreamFilters(User loggedInUser, String prefix,
+            String channelLabel, String projectLabel) throws ModulemdApiException {
+        ensureOrgAdmin(loggedInUser);
+
+        Channel channel = ChannelManager.lookupByLabelAndUser(channelLabel, loggedInUser);
```
Comment (cbbayburt) [inline java/code/src/com/redhat/rhn/frontend/xmlrpc/contentmgmt/ContentManagementHandler.java:668]: For proper error reporting, XMLRPC methods must throw a subclass of `FaultException` when something goes wrong. 

`ChannelManager.lookupByLabelAndUser` would throw a `LookupException` if the channel does not exist. You need to catch this and throw an `EntityNotExistsFaultException` instead.

File: java/code/src/com/redhat/rhn/frontend/xmlrpc/contentmgmt/ContentManagementHandler.java [hunk 3]
```diff
@@ -645,6 +651,39 @@ public ContentFilter createFilter(User loggedInUser, String name, String rule, S
...
+
+            for (ContentFilter createdFilter : createdFilters) {
+                attachedFilters.add(contentManager.attachFilter(projectLabel, createdFilter.getId(), loggedInUser));
+            }
+
+            return attachedFilters;
+        }
+        catch (EntityExistsException e) {
+            throw new EntityExistsFaultException(e);
+        }
```
Comment (cbbayburt) [inline java/code/src/com/redhat/rhn/frontend/xmlrpc/contentmgmt/ContentManagementHandler.java:684]: Same as above, here we need another catch block to catch the `EntityNotExistsException` (thrown from `attachFilter` if the project label does not exist) and throw an `EntityNotExistsFaultException`

**Review by cbbayburt — CHANGES_REQUESTED** [review]
Thanks for the PR @wombelix !
I commented a few suggestions about exception handling. Other than that, it looks good to me :+1:

File: java/code/src/com/redhat/rhn/frontend/xmlrpc/contentmgmt/ContentManagementHandler.java (hunk 1 above)
Comment (wombelix) [reply java/code/src/com/redhat/rhn/frontend/xmlrpc/contentmgmt/ContentManagementHandler.java:663]: Thanks @mcalmer @cbbayburt, totally forgot about that

File: java/code/src/com/redhat/rhn/frontend/xmlrpc/contentmgmt/ContentManagementHandler.java (hunk 2 above)
Comment (wombelix) [reply java/code/src/com/redhat/rhn/frontend/xmlrpc/contentmgmt/ContentManagementHandler.java]: Thanks, I tried to follow your suggestion, I moved that part into the existing try/catch block

File: java/code/src/com/redhat/rhn/frontend/xmlrpc/contentmgmt/ContentManagementHandler.java (hunk 3 above)
Comment (wombelix) [reply java/code/src/com/redhat/rhn/frontend/xmlrpc/contentmgmt/ContentManagementHandler.java:692]: I combined `LookupException` and `EntityNotExistsException` in the same catch block, not sure if that style is ok or if you prefer a separate catch block per exception.

**Comment by wombelix** [comment]
Didn't saw any issue related to this PR when running the Junit tests locally, unfortunately I can't access details of the failed `java_pgsql_tests` job https://ci.suse.de/job/uyuni-prs-java_pgsql_tests-pipeline/62650/

**Review by mcalmer — APPROVED** [review]

**Review by cbbayburt — APPROVED** [review]
Thanks for addressing all the comments @wombelix 
Looks all good now :+1:

**Comment by cbbayburt** [comment]
> Didn't saw any issue related to this PR when running the Junit tests locally, unfortunately I can't access details of the failed java_pgsql_tests job https://ci.suse.de/job/uyuni-prs-java_pgsql_tests-pipeline/62650/

I just retriggered the unit tests. Let's see after they're done. Sometimes the infrastructure may fail as well

**Comment by cbbayburt** [comment]
Let's wait for the remaining tests to run, then I'll merge it

**Pull request merged**